CMakeLists.txt

```diff
@@ -60,8 +60,22 @@ endif()
 
 enable_language(${SWIFT_ASM_DIALECT})
 
-# Use C++14.
-set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ standard to conform to")
+# Use C++17.
+set(SWIFT_MIN_CXX_STANDARD 17)
+
+# Unset CMAKE_CXX_STANDARD if it's too low and in the CMakeCache.txt
+if($CACHE{CMAKE_CXX_STANDARD} AND $CACHE{CMAKE_CXX_STANDARD} LESS ${SWIFT_MIN_CXX_STANDARD})
+  message(WARNING "Resetting cache value for CMAKE_CXX_STANDARD to ${SWIFT_MIN_CXX_STANDARD}")
+  unset(CMAKE_CXX_STANDARD CACHE)
+endif()
+
+# Allow manually specified CMAKE_CXX_STANDARD if it's greater than the minimum
+# required C++ version
+if(DEFINED CMAKE_CXX_STANDARD AND CMAKE_CXX_STANDARD LESS ${SWIFT_MIN_CXX_STANDARD})
+  message(FATAL_ERROR "Requested CMAKE_CXX_STANDARD=${CMAKE_CXX_STANDARD} which is less than the minimum C++ standard ${SWIFT_MIN_CXX_STANDARD}")
+endif()
+
+set(CMAKE_CXX_STANDARD ${SWIFT_MIN_CXX_STANDARD} CACHE STRING "C++ standard to conform to")
```

Comment:
Does this not need to be forced?

Reply:
`unset(CMAKE_CXX_STANDARD)` will override it. I didn't wan't to force it in case someone did `-DCMAKE_CXX_STANDARD=42` or had it set in some other way.

I think this is almost identical to how LLVM is handling the update now.

Reply:
* Correction: `unset` will remove it from the cache and let the `CACHE` var write it back with the new value.

Reply:
for reference: https://github.com/llvm/llvm-project/blob/a65678b35bbce31f09aebb2d1188c385b9dbac5e/llvm/CMakeLists.txt#L72-L86